sklearn/preprocessing/_data.py

```diff
@@ -1837,6 +1842,350 @@ def transform(self, X):
+
+        if self.knots in ['uniform', 'quantile']:
+            base_knots = self._get_base_knot_positions(
+                X, n_knots=self.n_knots, knots=self.knots)
+        else:
+            base_knots = check_array(self.knots)
+            if base_knots.shape[0] < 2:
+                raise ValueError("Number of knots, knots.shape[0], must be >= "
+                                 "2.")
+            elif base_knots.shape[1] != n_features:
+                raise ValueError("knots.shape[1] == n_features is violated.")
+            elif not np.all(np.diff(base_knots, axis=0) > 0):
+                raise ValueError("knots must be sorted without duplicates.")
+
+        if self.extrapolation not in ('error', 'constant', 'linear',
+                                      'continue'):
+            raise ValueError("extrapolation must be one of 'error', "
+                             "'constant', 'linear' or 'continue'.")
+
+        if not isinstance(self.include_bias, bool):
+            raise ValueError("include_bias must be bool.")
+
+        # number of knots for base interval
+        n_knots = base_knots.shape[0]
+        # number of splines basis functions
+        n_splines = n_knots + self.degree - 1
+        degree = self.degree
+        n_out = n_features * n_splines
+        # We have to add degree number of knots below, and degree number knots
+        # above the base knots in order to make the spline basis complete.
+        # Eilers & Marx in "Flexible smoothing with B-splines and  penalties"
+        # https://doi.org/10.1214/ss/1038425655 advice against repeating first
+        # and last knot several times, which would have inferior behaviour at
+        # boundaries if combined with a penalty (hence P-Spline). We follow
+        # this advice even if our splines are unpenalized.
+        # Meaning we do not:
+        # knots = np.r_[np.tile(base_knots.min(axis=0), reps=[degree, 1]),
+        #              base_knots,
+        #              np.tile(base_knots.max(axis=0), reps=[degree, 1])
+        #              ]
+        # Instead, we reuse the distance of the 2 fist/last knots.
+        dist_min = base_knots[1] - base_knots[0]
+        dist_max = base_knots[-1] - base_knots[-2]
+        knots = np.r_[linspace(base_knots[0] - degree * dist_min,
+                               base_knots[0] - dist_min, num=degree),
+                      base_knots,
+                      linspace(base_knots[-1] + dist_max,
+                               base_knots[-1] + degree * dist_max,
+                               num=degree)]
+
+        # With a diagonal coefficient matrix, we get back the spline basis
+        # elements, i.e. the design matrix of the spline.
+        # Note, BSpline appreciates C-contiguous arrays.
+        coef = np.eye(n_knots + self.degree - 1, dtype=np.float64)
+        extrapolate = self.extrapolation == "continue"
+        bsplines = [BSpline(knots[:, i], coef, self.degree,
```

Comment:
Since we constructed everything manually, can we take advantage of `construct_fast`?

```suggestion
        bsplines = [BSpline.construct_fast(knots[:, i], coef, self.degree,
```

Reply:
It seems so.